### PR TITLE
Tighten node repository locking

### PR DESCRIPTION
**Pull request by jonmv**
@hakonhall please review. Not sure I want to merge just yet.

### CODE REVIEW

**Comment by jonmv** [comment]
There are still several places where we read nodes, take decisions based on that, and then take the lock and perform an action. We could modify and use the `performOn` with a predicate argument, and a base list of nodes, to take the lock for each node, and perform the action only if the predicate still holds.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java [hunk 1]
```diff
@@ -880,6 +889,166 @@ private Node requireNode(String hostname) {
...
+                List<Node> nodes = new ArrayList<>(children.size() + 1);
+                nodes.addAll(children);
+                node.ifPresent(nodes::add);
+                mutexes = lockAndGetAll(nodes, timeout);
+                unallocatedLock = timeout.map(db::lockInactive).orElseGet(db::lockInactive);
+                RecursiveNodeMutexes recursive = new RecursiveNodeMutexes(hostname, mutexes, unallocatedLock);
+                Set<Node> freshChildren = recursive.children.stream().map(NodeMutex::node).collect(Collectors.toSet());
+                Optional<Node> freshNode = recursive.parent.map(NodeMutex::node);
+                if (children.equals(freshChildren) && node.equals(freshNode)) {
+                    // No new nodes have appeared, and none will now, so we have a consistent snapshot.
```
Comment (hakonhall) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java:922]: `lockAndGetAll(nodes, timeout)` will return a `mutexes` which is a subset of `nodes`.  The subset is possibly the same set as that denoted by `nodes`.  `recursive` will only return the children in `mutexes`.  Therefore `freshChildren` is a subset of `children`, and equal only if no children were removed.  This method doesn't handle the case where more children were added to the node before the unallocated lock were taken.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/persistence/CuratorDb.java [hunk 2]
```diff
@@ -116,10 +117,10 @@ public List<Node> addNodesInState(List<Node> nodes, Node.State expectedState, Ag
             curatorTransaction.add(CuratorOperations.create(nodePath(node).getAbsolute(), serialized));
         }
         transaction.onCommitted(() -> nodes.forEach(node -> log.log(Level.INFO, "Added " + node)));
-        return nodes;
+        return new ArrayList<>(nodes.asList());
```
[the remark this answers is not in the document]
Comment (jonmv) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/persistence/CuratorDb.java]: Yes, it broke some test code. I fixed that code instead now.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java [hunk 3]
```diff
@@ -880,6 +889,166 @@ private Node requireNode(String hostname) {
...
+                }
+
+                // If the first node is now earlier in lock order than some other locks we have, we need to close those and re-acquire them.
+                Node next = unlocked.pollFirst();
+                Set<NodeMutex> outOfOrder = locked.tailSet(new NodeMutex(next, () -> { }), false);
+                NodeMutexes.close(outOfOrder.iterator());
+                for (NodeMutex node : outOfOrder) unlocked.add(node.node());
+                outOfOrder.clear();
+
+                Mutex lock = lock(next, timeout);
```
[the remark this answers is not in the document]
Comment (jonmv) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java:984]: Yes, I thought about this as well, but forgot to actually do something about it. What's a reasonable aggregate timeout, then? `sqrt(nodes.size()) * timeout`?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java (hunk 1 above)
Comment (jonmv) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java:922]: Oops, great catch. Yes, I "optimised away a read" there 🤦

Comment (jonmv) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java:922]: Restoring to the original additional `list().childrenOf(hostname)`.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java [hunk 4]
```diff
@@ -726,35 +739,33 @@ private List<Node> performOn(Predicate<Node> filter, BiFunction<Node, Mutex, Nod
...
-                unallocatedNodes.add(node);
-        }
+    public List<Node> performOn(NodeList nodes, Predicate<Node> filter, BiFunction<Node, Mutex, Node> action) {
+        List<Node> resultingNodes = new ArrayList<>();
+        nodes.stream().collect(groupingBy(Nodes::applicationIdForLock))
+             .forEach((applicationId, nodeList) -> { // Grouped only to reduce number of lock acquire/release cycles.
+                 try (NodeMutexes locked = lockAndGetAll(nodeList, Optional.empty())) {
+                     for (NodeMutex node : locked.nodes())
+                         if (filter.test(node.node()))
+                             resultingNodes.add(action.apply(node.node(), node));
```
[the remark this answers is not in the document]
Comment (jonmv) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java:753]: They're all just writes to ZK, so I think it's best this way.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonmv) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java]: Well, I'll leave that to the users. Changed to a TimeBudget now.

**Review by hakonhall — APPROVED** [review]

**Pull request merged**